compositor_common/src/scene/constraints.rs

```diff
@@ -0,0 +1,57 @@
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    error::UnsatisfiedConstraintsError,
+    scene::{NodeId, SceneSpec},
+};
+
+use self::input_count::InputsCountConstraint;
+
+use super::NodeSpec;
+
+pub mod input_count;
+
+// TODO validate constraints aren't self-contradictory
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct NodeConstraints(pub(crate) Vec<Constraint>);
+
+impl NodeConstraints {
+    pub fn check(
```

Comment:
`check` instead of `validate`, because `validate` can mean that constraints spec validation